clients/src/main/java/org/apache/kafka/common/requests/AbstractRequest.java

```diff
@@ -93,7 +93,13 @@ public short version() {
     }
 
     public Send toSend(String destination, RequestHeader header) {
-        return new NetworkSend(destination, serialize(header));
+        // For UpdateMetadataRequest, the toSend method on the same object will be called many times, each time with a different destination
+        // value and a header containing a different correlation id.
+        ByteBuffer headerBuffer = serializeStruct(header.toStruct());
+        if (bodyBuffer == null) {
```

Comment:
Question: 

Could there be multiple threads calling the `toSend` method? If yes, we would need to have some atomic reference to wrap on the bodyBuffer variable right?

Reply:
I had the wrong impression that the controller is single-threaded. But actually there is a separate RequestSendThread for each broker. So yes the access needs to be synchronized. 
BTW, this PR is using the obsolete branch. I'm closing it, please send your comments to the updated one https://github.com/linkedin/kafka/pull/75.